### PR TITLE
Images importation

### LINKED ISSUE #176: Manage images of molecules for questions
- [ ] Serve the image with the server
- [ ] Cache the image in the client

### PR DESCRIPTION
**Pull request by FrancoisPog**
## Import the images of molecules 
closes #176 

- [x] Create images analyzer 
- [x] Improve multer middleware to manage multiple files
- [x] Improve admin view to send multiple files 
- [x] Create the route to analyze & import images 
- [x] Update database to add `mo_image` attribute to molecules
- [x] Set route to download imported images in an archive
- [x] Serve images statically
- [x] Docs
- [x] Tests

### CODE REVIEW

File: server/global/images_importation/imagesImporter.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { removeExtension } from "../Files.js";
+
+import { addErrorTitle } from "../Logger.js";
+
+import { normalizeDCI } from "../molecules_importation/moleculesAnalyzer.js";
+
+import { isFormatValid } from "./imagesAnayzer.js";
+
+/**
+ * Update the molecule image in database, folloxing the given filenames
```
Comment (valentinperignon) [inline server/global/images_importation/imagesImporter.js:14]:
```suggestion
 * Update the molecule image in database, following the given filenames
```

File: server/test/images_importation/imagesAnalyzer.test.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    ],
+    warnings: [
+      { code: ImagesAnalyzerWarning.DUPLICATE_IMAGES, count: 2 },
+      { code: ImagesAnalyzerWarning.UNKNOWN_MOLECULES, count: 0 },
+      { code: ImagesAnalyzerWarning.INVALID_MOLECULE, count: 0 },
+      { code: ImagesAnalyzerWarning.BAD_FORMAT, count: 0 },
+    ],
+  },
+  {
+    filenames: ["amantadine.tex", "lamivudine.odt", "tinidazole", "pyrantel.jpEg ", "   "],
```
Comment (valentinperignon) [inline server/test/images_importation/imagesAnalyzer.test.js:38]: We can't import tech files? 😱

**Review by valentinperignon — APPROVED** [review]
LGTM!
Really great work (even if you can do much better, only 84 files...)

File: server/test/images_importation/imagesAnalyzer.test.js (hunk 2 above)
Comment (FrancoisPog) [reply server/test/images_importation/imagesAnalyzer.test.js:38]: I am really sorry ...

File: server/global/Files.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      .catch((error) => {
+        if (error.code === "ENOENT") {
+          resolve([]);
+        }
+        reject(addErrorTitle(error, "Can't read the directory"));
+      });
+  });
+}
+
+/**
```
Comment (nathanaelhoun) [inline server/global/Files.js:76]: FYI, nodeJS' `path` module [has a method for this](https://nodejs.org/api/path.html#path_path_parse_path).

File: server/global/images_importation/imagesAnayzer.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import { addErrorTitle } from "../Logger.js";
+
+import {
+  normalizeDCI,
+  getInvalidNormalizedDci,
+} from "../molecules_importation/moleculesAnalyzer.js";
+
+const VALID_FORMATS = ["jpeg", "jpg", "svg", "png"];
+export const isFormatValid = (filename) =>
+  new RegExp(`\\.${VALID_FORMATS.join("|")}\\s*$`, "i").test(filename);
```
Comment (nathanaelhoun) [inline server/global/images_importation/imagesAnayzer.js:12]: You could store the regex in a constant, as it we won't regeneate it each time the function is called

File: server/global/images_importation/imagesAnayzer.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+/**
+ * Analyze images filenames
+ * @param {string[]} filenames The list of images filenames
+ * @returns {Promise<ImagesAnalyzerWarning[]>} The warnings list
+ */
+export function analyseImagesFilenames(filenames) {
+  return new Promise((resolve, reject) => {
+    const warnings = [];
+
+    const invalidFormats = filenames.filter((f) => !isFormatValid(f));
```
Comment (nathanaelhoun) [inline server/global/images_importation/imagesAnayzer.js:23]: I think we could also rely on the [MIME type](https://developer.mozilla.org/fr/docs/Web/HTTP/Basics_of_HTTP/MIME_Types) and not only on the file extension, as the former is more reliable.

File: server/global/images_importation/imagesImporter.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import mysql from "mysql";
+
+import { queryPromise } from "../../db/database.js";
+
+import { removeExtension } from "../Files.js";
+
+import { addErrorTitle } from "../Logger.js";
+
+import { normalizeDCI } from "../molecules_importation/moleculesAnalyzer.js";
```
Comment (nathanaelhoun) [inline server/global/images_importation/imagesImporter.js:9]: Useless newlines?

File: server/global/molecules_importation/moleculesAnalyzer.js [hunk 7]
```diff
@@ -173,18 +181,39 @@ function getNonValidNumberValue(molecules) {
...
-export class AnalyzerWarning {
+/**
+ * Normalize a molecule DCI
+ * @param {string} dci
+ */
+export const normalizeDCI = (dci) =>
+  String(dci)
+    .trim()
+    .normalize("NFD")
+    .replace(/[\u0300-\u036f]/g, "")
```
Comment (nathanaelhoun) [inline server/global/molecules_importation/moleculesAnalyzer.js:192]: Can a comment explain what this regexp is doing?

File: server/middlewares/multer.middleware.js [hunk 8]
```diff
@@ -2,32 +2,24 @@ import fs from "fs/promises";
...
  * Multer configuration
  */
 const storage = multer.diskStorage({
   filename: (req, file, callback) => {
-    const filename = String(Date.now());
-    const extension = MIME_TYPES[file.mimetype];
-
-    req.body._uploadedFileName = filename;
-    req.body._uploadedFileExtension = extension;
+    const extension = file.originalname.split(".").slice(1).pop();
```
Comment (nathanaelhoun) [inline server/middlewares/multer.middleware.js:15]: Same as above with node's `path` module

File: server/middlewares/multer.middleware.js [hunk 9]
```diff
@@ -2,32 +2,24 @@ import fs from "fs/promises";
...
+    const filename = `${Date.now()}.${uuidv4()}.${extension}`;
```
Comment (nathanaelhoun) [inline server/middlewares/multer.middleware.js:16]: It's great to use uuids :+1: Can we keep the original filename (for example the dci) in the final name to figure out what the file is with it's name?

File: server/package.json [hunk 10]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "type": "module",
   "scripts": {
-    "test": "NODE_ENV=test mocha ./test/* --exit",
+    "test": "NODE_ENV=test mocha \"test/**/*.test.js\" --exit",
```
Comment (nathanaelhoun) [inline server/package.json:8]: Curiosity: why this change?

File: server/package.json [hunk 11]
```diff
@@ -50,8 +56,18 @@
...
+          "test/**"
+        ],
+        "rules": {
+          "promise/catch-or-return": "off"
+        }
+      }
+    ],
     "parserOptions": {
-      "ecmaVersion": 2017,
+      "ecmaVersion": 2020,
```
Comment (nathanaelhoun) [inline server/package.json:70]: Oh man we are so in!

File: server/test/images_importation/importation_route/importImages.test.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+  });
+});
+
+/**
+ * Send request to import files of a given directory
+ * @param {string} dir The directory containing files to send
+ * @param {string} confirmed Tell if the importation is confirmed
+ * @param {string} token The user token
+ */
+function importImages(dir, confirmed = "", token = "") {
```
Comment (nathanaelhoun) [inline server/test/images_importation/importation_route/importImages.test.js:138]:
```suggestion
function importImagesViaAPI(dir, confirmed = "", token = "") {
```

**Review by nathanaelhoun — COMMENTED** [review]
I will test this tomorrow, already some comments and questions on the (great) work :rocket: 

(please next time don't bother us with such a small change /s)

File: server/controllers/imagesImporter.js [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+                      imported: true,
+                    });
+
+                    deleteFiles(
+                      ...req.files
+                        .filter((f) => !imported.includes(f.originalname))
+                        .map((f) => f.path)
+                    ).catch(Logger.error);
+                  })
+                  .catch((error) => sendServerError(error, "Can't move images"))
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/controllers/imagesImporter.js:105]: Error messages won't be displayed to users but only in logs so we should keep english for them. For message for users, an other PR will traduce them.

File: server/global/images_importation/imagesImporter.js (hunk 6 above)
Comment (FrancoisPog) [reply server/global/images_importation/imagesImporter.js:9]: Eslint did it on his own 🤷‍♂️

File: server/middlewares/multer.middleware.js (hunk 9 above)
Comment (FrancoisPog) [reply server/middlewares/multer.middleware.js:16]: Not necessary because the middleware stores all the information about each file, so in the controllers we can retrieve the original name (among other information).

File: server/package.json (hunk 10 above)
Comment (FrancoisPog) [reply server/package.json:8]: I don't know why, but the first one doesn't fully browse the tests directory, so I change that to run all tests, even those in sub-subfolders.

File: server/global/images_importation/imagesAnayzer.js (hunk 5 above)
Comment (FrancoisPog) [reply server/global/images_importation/imagesAnayzer.js:23]: I did this initially, but in reality there are some non-standard MIME TYPE that we cannot handle, for example for CSV files the standard is `text/csv`, but also `application/vnd.ms-excel`, and surely others. It is therefore difficult for us to verify this, at the risk of refusing a valid file.
For our use, only admins will import the files, so we can trust them. In addition, they will receive an import manual explaining how to do it correctly.

File: server/global/images_importation/imagesAnayzer.js (hunk 4 above)
Comment (FrancoisPog) [reply server/global/images_importation/imagesAnayzer.js:12]: done! 👍

File: server/global/molecules_importation/moleculesAnalyzer.js (hunk 7 above)
Comment (FrancoisPog) [reply server/global/molecules_importation/moleculesAnalyzer.js:192]: done 👍

File: server/test/images_importation/importation_route/importImages.test.js (hunk 12 above)
Comment (FrancoisPog) [reply server/test/images_importation/importation_route/importImages.test.js:138]: done 👍

File: server/middlewares/multer.middleware.js (hunk 8 above)
Comment (FrancoisPog) [reply server/middlewares/multer.middleware.js:15]: done 👍

File: server/global/Files.js (hunk 3 above)
Comment (FrancoisPog) [reply server/global/Files.js:76]: done 👍

**Review by nathanaelhoun — APPROVED** [review]
Really awesome job!
